autopayroll-CordaService/workflows/src/main/kotlin/com/autopayroll/flows/AutoPaymentTrigger.kt

```diff
@@ -0,0 +1,46 @@
+package com.autopayroll.flows
+
+import com.autopayroll.states.PaymentRequestState
+import net.corda.core.contracts.StateAndRef
+import net.corda.core.identity.CordaX500Name
+import net.corda.core.identity.Party
+import net.corda.core.node.AppServiceHub
+import net.corda.core.node.services.CordaService
+import net.corda.core.node.services.Vault
+import net.corda.core.node.services.trackBy
+import net.corda.core.serialization.SingletonSerializeAsToken
+import net.corda.core.utilities.loggerFor
+import java.util.concurrent.Executor
+import java.util.concurrent.Executors
+
+@CordaService
+class AutoPaymentTrigger(private val serviceHub: AppServiceHub) : SingletonSerializeAsToken() {
+
+    private companion object {
+        val log = loggerFor<AutoPaymentTrigger>()
+        val executor: Executor = Executors.newFixedThreadPool(8)!!
+    }
+
+    init {
+        directPayment()
+        log.info("Tracking new Payment Request")
+    }
+
+    private fun directPayment() {
+        val ourIdentity = ourIdentity()
+        serviceHub.vaultService.trackBy<PaymentRequestState>().updates.subscribe {
+            update: Vault.Update<PaymentRequestState> -> update.produced.forEach{
+            message: StateAndRef<PaymentRequestState> ->
+                val state = message.state
+                if (ourIdentity == serviceHub.networkMapCache.getPeerByLegalName(CordaX500Name("BankOperator", "Toronto", "CA"))!!) {
```

Comment:
This is interesting. You've had to do this because every node has the service but you don't want people to automatically pay others unless they're the bank right? I imagine the way you'd actually go about doing this would be to package the bit of the app that defines the service separately and only deploy it in places that should actually run it. I think it's worth a comment here explaining what's going on. You could also try splitting the app up in this way.

Reply:
This would actually be quite an important part of the sample as this restriction comes up surprisingly frequently and a good solution is not widely published and explained.

Reply:
In the real production scenario, it might be a good practice to split the service and the rest of the flows, and compile them into two jars. But I think in this specific sample, it will complicate the corDapp.